File: applications/llm/tio/src/input/http.rs
// SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

use triton_distributed_llm::http::service::{discovery, service_v2};
use triton_distributed_runtime::{DistributedRuntime, Runtime};

use crate::EngineConfig;

/// Build and run an HTTP service
pub async fn run(
    runtime: Runtime,
    http_port: u16,
    engine_config: EngineConfig,
) -> anyhow::Result<()> {
    let http_service = service_v2::HttpService::builder()
        .port(http_port)
        .enable_chat_endpoints(true)
        .enable_cmpl_endpoints(true)
        .build()?;
    match engine_config {
        EngineConfig::Dynamic(client) => {
            let service_name = client.path();
            let distributed_runtime = DistributedRuntime::from_settings(runtime.clone()).await?;
            // Listen for models registering themselves in etcd, add them to HTTP service
            let state = Arc::new(discovery::ModelWatchState {
                prefix: service_name.clone(),
                manager: http_service.model_manager().clone(),
                drt: distributed_runtime.clone(),
            });
            let etcd_client = distributed_runtime.etcd_client();
            let models_watcher = etcd_client.kv_get_and_watch_prefix(service_name).await?;
            let (_prefix, _watcher, receiver) = models_watcher.dissolve();
            let _watcher_task = tokio::spawn(discovery::model_watcher(state, receiver));
        }
        EngineConfig::StaticFull {
            service_name,
            engine,
            ..
        } => {
            http_service
                .model_manager()
                .add_chat_completions_model(&service_name, engine)?;
        }
    }
    http_service.run(runtime.primary_token()).await
}

File: deploy/dynamo/operator/api/dynamo/yatai-client/http.go
/*
 * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package yataiclient

import (
	"context"
	"crypto/tls"
	"fmt"
	"time"

	"resty.dev/v3"
)

var defaultClient *resty.Client

func GetDefaultClient() *resty.Client {
	if defaultClient == nil {
		defaultClient = resty.New().
			SetTimeout(90*time.Second).
			SetRetryCount(3).
			SetRetryWaitTime(2*time.Second).
			SetRetryMaxWaitTime(10*time.Second).
			SetHeader("Content-Type", "application/json").
			SetTLSClientConfig(&tls.Config{InsecureSkipVerify: true}) // Optional: mirrors your custom transport
	}
	return defaultClient
}

func DoJsonRequest(ctx context.Context, method string, url string, headers map[string]string, query map[string]string, payload interface{}, result interface{}, timeout *time.Duration) (int, error) {
	client := GetDefaultClient()

	if timeout != nil {
		client.SetTimeout(*timeout)
	}

	req := client.R().
		SetContext(ctx).
		SetBody(payload).
		SetResult(result).
		SetHeaders(headers).
		SetQueryParams(query)

	var resp *resty.Response
	var err error

	switch method {
	case "GET":
		resp, err = req.Get(url)
	case "POST":
		resp, err = req.Post(url)
	case "PUT":
		resp, err = req.Put(url)
	case "DELETE":
		resp, err = req.Delete(url)
	case "PATCH":
		resp, err = req.Patch(url)
	default:
		return 0, fmt.Errorf("unsupported method: %s", method)
	}

	if err != nil {
		return 0, fmt.Errorf("request error: %w", err)
	}

	if resp.IsError() {
		return resp.StatusCode(), fmt.Errorf("http %s %s failed with status %d: %s", method, url, resp.StatusCode(), resp.String())
	}

	return resp.StatusCode(), nil
}
